### PR TITLE
feat: concurrent alt-da requests

### PR DESCRIPTION
**Pull request by samlaf**
This is the simplest possible implementation to fix the non-blocking da requests. We build off of @karlb's https://github.com/celo-org/optimism/pull/213. We use an errgroup for da requests, with its own limit to the number of concurrent goroutines. See https://hackmd.io/@samlaf/op-batcher-concurrent-altda-requests-design-doc for more details.

### Testing
We added a basic [e2e test](https://github.com/Layr-Labs/optimism/pull/7/files#diff-8bc46de6e90dc6a2a1773e2a90eca15e631d5b7e062fc2f282371082faa30e52R1406) by using a fakeDaServer that sleeps for 5 seconds before returning commitments, in order to simulate slow EigenDA responses.

We also tested this manually by running the devnet and submitting large txs and observing the behavior.

### Open Questions 
(see https://hackmd.io/@samlaf/op-batcher-concurrent-altda-requests-design-doc#Open-questions for working links)
1. How to differentiate between stops and shutdowns for da requests? txmgr uses a separate txmgr.Close() function to quickly kill pending txs during shutdown. We might want to have a similar mechanism for DA requests? But in the meantime, we’ve opted to cancel all pending DA requests for both closes and shutdowns, by using l.shutdownCtx (see here and comment above) instead of the currently used l.killCtx . This seems fine since the txData is pushed back into the channel on error, and would be resubmitted on a later start().
2. errgroups for alt-da and eth-da are different: should we combine them into a single goroutine that does both calls serially. This would also mean combining their max concurrent goroutine flags/parameters (MaxPendingTransactions and MaxConcurrentDARequests).
3. How do we feel about changing the MaxL1TxSizeBytes flag/param to MaxFrameSizeBytes to better represent its dual use in the case of alt-da, where the actual l1 tx size is the size of the commitment, not of the frame?

### CODE REVIEW

**Comment by tynes** [comment]
This is extremely well thought through, thank you @samlaf for this! @sebastianst is the main owner of the batcher right now

File: op-alt-da/damock.go
Comment (tynes) [inline op-alt-da/damock.go:127]: Isn't this going to cause tests to take 5 seconds times the number of times this is called extra amount of time to run?

File: op-batcher/batcher/driver.go
Comment (tynes) [inline op-batcher/batcher/driver.go]: small nit: might be better to have conditional logic here with the logging based on whether alt da is on or not

**Comment by tynes** [comment]
@roberto-bayardo could also be a good person to help with review here

**Comment by tynes** [comment]
Generally looks good to me although I do not own this code

File: op-batcher/batcher/driver.go
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: Please extract this into a separate method like `sendAltDATransaction`.

File: op-batcher/batcher/driver.go
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: Agree, would be good to put this into a `if l.Config.UseAltDA` branch. Plus, put the `daGroup.Wait()` call into a ` if err := daGroup.Wait(); err != nil { ...` branch.

File: op-e2e/system_test.go
Comment (sebastianst) [inline op-e2e/system_test.go]: why does it need to be 10 blocks, not just e.g. 2?

File: op-e2e/system_test.go
Comment (sebastianst) [inline op-e2e/system_test.go]: why wouldn't waiting for 2*L1BlockTime be sufficient?

File: op-e2e/system_test.go
Comment (sebastianst) [inline op-e2e/system_test.go]: Note that there are possibly other services in the background sending transactions, like the proposer or challenger. I think it would be better to only count the transactions having the batcher as sender.

File: op-alt-da/damock.go
Comment (sebastianst) [reply op-alt-da/damock.go:127]: I understand that because of the newly introduced concurrency in sending DA transactions, multiple submit routines would wait those 5 secs in parallel.

**Review by sebastianst — CHANGES_REQUESTED** [review]

File: op-alt-da/damock.go
Comment (samlaf) [reply op-alt-da/damock.go:127]: Correct, this is only used in the concurrent da requests test.

File: op-batcher/batcher/driver.go
Comment (samlaf) [reply op-batcher/batcher/driver.go]: [2efafab](https://github.com/ethereum-optimism/optimism/pull/11698/commits/2efafab0ad0e0208d7fe853fd840259f24c4e856)

@sebastianst I just copied the logic of the quque's waitgroup, which is also [throwing away its error](https://github.com/ethereum-optimism/optimism/blob/42085ce945bc727707b9d51c8e90e142b05f721e/op-service/txmgr/queue.go#L51). Do you prefer we catch them and log something?

File: op-batcher/batcher/driver.go
Comment (samlaf) [reply op-batcher/batcher/driver.go]: [f0e6886](https://github.com/ethereum-optimism/optimism/pull/11698/commits/f0e6886a0f33a3103c8c825f017793ceceb363e9)
Named the function `publishToAltDAAndL1`, but open to renaming.

File: op-e2e/system_test.go
Comment (samlaf) [reply op-e2e/system_test.go]: That's a good question. I just copied the TestBatcherMultiTx test above it, which [uses 10](https://github.com/ethereum-optimism/optimism/blob/42085ce945bc727707b9d51c8e90e142b05f721e/op-e2e/system_test.go#L1375).

File: op-e2e/system_test.go
Comment (samlaf) [reply op-e2e/system_test.go]: Again just copied TestBatcherMultiTx along with its constants. Shall I change both?

File: op-e2e/system_test.go
Comment (samlaf) [reply op-e2e/system_test.go]: [1185bf8](https://github.com/ethereum-optimism/optimism/pull/11698/commits/1185bf87cfebab54f680984010275303cfbff8a2)
Shall I also change the other test?

File: op-batcher/batcher/driver.go
Comment (axelKingsley) [inline op-batcher/batcher/driver.go:622]: great abstraction/refactor :)

File: op-batcher/batcher/driver.go
Comment (axelKingsley) [inline op-batcher/batcher/driver.go:603]: I'm trying to think about what might happen if a submission to DA is particularly slow, and the ones behind it are fast.

I'm assuming that all later transactions would be blocked for nonce reasons until the first transaction finishes, and then all the others would unblock, does that sound right? For AltDAs that can handle out of order `SetInput` (like an S3 bucket or something), only the commitment to L1 would be blocked by previous transactions. But some AltDAs would force serialization.

In the cases where AltDA have forced serialization on submission (like a nonce), having one goroutine per transaction doesn't fully parallelize the work, it just moves it out of the main thread, correct?

Comment (samlaf) [reply op-batcher/batcher/driver.go:603]: > I'm trying to think about what might happen if a submission to DA is particularly slow, and the ones behind it are fast.

Do you have an example in mind? I can't think of a DA layer where the delta would be meaningful. Unless you're thinking one blob is huge and the other is small for eg?

> I'm assuming that all later transactions would be blocked for nonce reasons until the first transaction finishes, and then all the others would unblock, does that sound right?

Not quite. I added the "open questions" section in the PR description. But the idea right now is that both da errgroup and txmgr errgroup are independent, and hence can have a different goroutine limit. This can be useful for eg if you want to limit eth txs to 5 in parallel (because more than that leads to nonce problems or whatever.. just an eg), whereas you need 100 da goroutines in parallel because they each take 10 mins. This is fine, and the da goroutines will simply block on the txmgr errorgroup to spawn a new txmgr goroutine. DA requests that are faster can also overtake the otherones and end up having their commitment sent to L1 earlier.

> But some AltDAs would force serialization.

Do you have an example in mind? I don't see why this would ever be needed/useful. In my mind, a DA layer should purely be for DA. Ordering should be done at the sequencing layer. Otherwise you run into some weird cases where DA layer could extract MEV, as one example.

> In the cases where AltDA have forced serialization on submission (like a nonce), having one goroutine per transaction doesn't fully parallelize the work, it just moves it out of the main thread, correct?

Still not super sure what you mean by this as I explained above. But let's say the DA layer for some reason requires a nonce to be submitted. You can always use the da server as that serialization coordinator, and have that as a middleware that adds a nonce. So seems like things would be fine even in this model?

Comment (axelKingsley) [reply op-batcher/batcher/driver.go:603]: For examples of a submission which is slow while others are fast: Yeah, data size is a good example, or even just some sort of network issue that disproportionately affects one transaction submission over others.

For an example of a DA with forced serialization: I don't have a concrete example in mind, and I agree that a DA *should* simply serve DA. So maybe there's no issue because I'm imagining a mode of operation that is unrealistic.

I think you're right that things are fine in this model, thanks for thinking about it with me :)

File: op-batcher/batcher/driver.go
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: Why not error log if this returns an error?

File: op-e2e/system_test.go
Comment (sebastianst) [reply op-e2e/system_test.go]: if it works, that'd be great, thank you!

File: op-e2e/system_test.go
Comment (sebastianst) [reply op-e2e/system_test.go]: That'd be great 🙏🏻  Maybe there's also an opportunity to extract the code that counts the number of txs by a certain sender within a block range into a common function, to avoid duplication.

File: op-e2e/system_test.go
Comment (sebastianst) [reply op-e2e/system_test.go]: Let's try a lower number like 3, here and also for the other test if it works. If it still works, this speeds up the test(s) and saves on CI time.

File: op-batcher/batcher/driver.go
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: thank you!

File: op-batcher/batcher/driver.go
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: I think it makse sense to error-log in case of error. No idea why it was thrown away before.

Comment (samlaf) [reply op-batcher/batcher/driver.go]: [3d31b9a](https://github.com/ethereum-optimism/optimism/pull/11698/commits/3d31b9a5d0fafc414d55252fe2823ee33dcf64ce)

File: op-batcher/batcher/driver.go
Comment (samlaf) [reply op-batcher/batcher/driver.go]: [3d31b9a](https://github.com/ethereum-optimism/optimism/pull/11698/commits/3d31b9a5d0fafc414d55252fe2823ee33dcf64ce)

File: op-e2e/system_test.go
Comment (samlaf) [reply op-e2e/system_test.go]: [d4362f3](https://github.com/ethereum-optimism/optimism/pull/11698/commits/d4362f34111c878a7738419164c8cf54afbcc792)

File: op-e2e/system_test.go
Comment (samlaf) [reply op-e2e/system_test.go]: [d4362f3](https://github.com/ethereum-optimism/optimism/pull/11698/commits/d4362f34111c878a7738419164c8cf54afbcc792)

File: op-e2e/system_test.go
Comment (samlaf) [reply op-e2e/system_test.go]: [4a43bd0](https://github.com/ethereum-optimism/optimism/pull/11698/commits/4a43bd04b5ce1d32cb3a48d8bf77602b4f426880)

**Comment by samlaf** [comment]
@sebastianst just realized I was not initializing `MaxConcurrentDARequests`... :|
Fixed in my last commit: [5e519ca](https://github.com/ethereum-optimism/optimism/pull/11698/commits/5e519ca9dc22bc9cea0469455b9dd1ed4a1f703a)

**Review by sebastianst — APPROVED** [review]

**Pull request merged**
